pkg/services/govmomi/service.go

```diff
@@ -273,6 +279,18 @@ func (vms *VMService) reconcilePowerState(ctx *virtualMachineContext) (bool, err
 	}
 }
 
+func (vms *VMService) reconcileTags(ctx *virtualMachineContext) error {
+	tagManager := tags.NewManager(ctx.Session.RestClient)
+
+	for _, tagName := range ctx.VSphereVM.Spec.Tags {
+		if err := tagManager.AttachTag(ctx, tagName, ctx.Ref); err != nil {
```

Comment:
I recommend some form of Client-side rate limiter for these REST calls, see this note:

> As an additional consideration, to avoid denial-of-service attacks to the tagging service, there is a limit of
360 tagging-related operations per second. To detect if you are hitting this limit, look for messages like “Request rejected due to high request rate. Try again later.” in the vapi-endpoint log file (/var/log/vmware/vapi/endpoint/endpoint.log). If your creation code is exceeding this limit, we recommend inserting a pause of 2 ms or more between tag-creation calls.

Also, I'll try to implement the batch API for more efficient (time, resources) tagging in govmomi soon.